### ISSUE TITLE
The reconstructed image is blurry

### ISSUE THREAD

**Issue by QijingGJ** [LemuelPuglisi/BrLP#10]
Hello, I tried train_autoencoder.py using my own data, but the reconstructed image was very blurry and many details were lost. What should I do? Is it due to insufficient training rounds?

Thank you very much for your help

![Image](https://github.com/user-attachments/assets/cca6884b-7fa4-46d3-a0a4-5cfde092bfea)

![Image](https://github.com/user-attachments/assets/6d3da7f9-f834-4001-92b0-302f6a7ca2ef)

**Comment by LemuelPuglisi** [comment]
Hi! 

The VAE loss in BrLP is composed of 4 terms (reconstruction, perceptual, adversarial, and KL regularisation). Could you please show me the training curves for all of them? 

Best regards, 
Lemuel

**Comment by QijingGJ** [comment]
Hello, this is the entire loss curve. Thank you very much for your help

> ![Image](https://github.com/user-attachments/assets/30c3eb88-5b90-4f39-863d-4387bfd320c1)
> 
> ![Image](https://github.com/user-attachments/assets/9294c22f-fdbb-4db5-9344-a850da2b56aa)

**Comment by LemuelPuglisi** [comment]
I suggest keeping training the network for more epochs. Also, try to keep the perceptual, adversarial and reconstruction losses to the same scale by modifying their weights.

**Comment by QijingGJ** [comment]
Thank you for your patient explanation. My data is a 2D cardiac MRI using AutoencoderKL spatial_dims=2. Will this have any impact

> I suggest keeping training the network for more epochs. Also, try to keep the perceptual, adversarial and reconstruction losses to the same scale by modifying their weights.

**Comment by LemuelPuglisi** [comment]
No problem! I think it's fine to use 2D images—just make sure to adapt the rest of the BrLP pipeline accordingly by adjusting the input dimensions for the other networks as well. Let me know if you have any questions!

**Comment by QijingGJ** [comment]
> No problem! I think it's fine to use 2D images—just make sure to adapt the rest of the BrLP pipeline accordingly by adjusting the input dimensions for the other networks as well. Let me know if you have any questions!

Thank you very much for your help.  I have successfully obtained a clear reconstructed image.  Your work is very interesting and amazing.  But I have two other doubts when running train_diffusion_unet.py, and I am very eager to get your reply.

（1）The first doubt is whether the calculation of scale_factor uses only one sample.  Why not all samples？

with torch.no_grad():
with autocast(enabled=True):
z = trainset[0]['latent']
scale_factor = 1 / torch.std(z)

（2）The second doubt is, what is the dimension of batch['context']?  Is it (batch_size, sequence_length, embedding_dim)?  Before entering the LDM through cross-attention, does the context need additional modules for embedding?

I'm really looking forward to your reply.  Best wishes.

**Comment by AlexNmSED** [comment]
> > No problem! I think it's fine to use 2D images—just make sure to adapt the rest of the BrLP pipeline accordingly by adjusting the input dimensions for the other networks as well. Let me know if you have any questions!
> 
> Thank you very much for your help. I have successfully obtained a clear reconstructed image. Your work is very interesting and amazing. But I have two other doubts when running train_diffusion_unet.py, and I am very eager to get your reply.
> 
> （1）The first doubt is whether the calculation of scale_factor uses only one sample. Why not all samples？
> 
> with torch.no_grad(): with autocast(enabled=True): z = trainset[0]['latent'] scale_factor = 1 / torch.std(z)
> 
> （2）The second doubt is, what is the dimension of batch['context']? Is it (batch_size, sequence_length, embedding_dim)? Before entering the LDM through cross-attention, does the context need additional modules for embedding?
> 
> I'm really looking forward to your reply. Best wishes.

Hello, I'm a bit curious about your solution — what caused the previous issue?  Thank you.

**Comment by QijingGJ** [comment]
> > > No problem! I think it's fine to use 2D images—just make sure to adapt the rest of the BrLP pipeline accordingly by adjusting the input dimensions for the other networks as well. Let me know if you have any questions!
> > 
> > 
> > Thank you very much for your help. I have successfully obtained a clear reconstructed image. Your work is very interesting and amazing. But I have two other doubts when running train_diffusion_unet.py, and I am very eager to get your reply.
> > （1）The first doubt is whether the calculation of scale_factor uses only one sample. Why not all samples？
> > with torch.no_grad(): with autocast(enabled=True): z = trainset[0]['latent'] scale_factor = 1 / torch.std(z)
> > （2）The second doubt is, what is the dimension of batch['context']? Is it (batch_size, sequence_length, embedding_dim)? Before entering the LDM through cross-attention, does the context need additional modules for embedding?
> > I'm really looking forward to your reply. Best wishes.
> 
> Hello, I'm a bit curious about your solution — what caused the previous issue? Thank you.

I increased the training epochs and adjusted the weight of the loss, and then it worked

**Comment by AlexNmSED** [comment]
> > > > No problem! I think it's fine to use 2D images—just make sure to adapt the rest of the BrLP pipeline accordingly by adjusting the input dimensions for the other networks as well. Let me know if you have any questions!
> > > 
> > > 
> > > Thank you very much for your help. I have successfully obtained a clear reconstructed image. Your work is very interesting and amazing. But I have two other doubts when running train_diffusion_unet.py, and I am very eager to get your reply.
> > > （1）The first doubt is whether the calculation of scale_factor uses only one sample. Why not all samples？
> > > with torch.no_grad(): with autocast(enabled=True): z = trainset[0]['latent'] scale_factor = 1 / torch.std(z)
> > > （2）The second doubt is, what is the dimension of batch['context']? Is it (batch_size, sequence_length, embedding_dim)? Before entering the LDM through cross-attention, does the context need additional modules for embedding?
> > > I'm really looking forward to your reply. Best wishes.
> > 
> > 
> > Hello, I'm a bit curious about your solution — what caused the previous issue? Thank you.
> 
> I increased the training epochs and adjusted the weight of the loss, and then it worked

Thank you for your reply.

**Comment by LemuelPuglisi** [comment]
Hi! I'm glad that you solved your issue!
 
> （1）The first doubt is whether the calculation of scale_factor uses only one sample. Why not all samples？

Read this section from the LDM paper https://arxiv.org/pdf/2112.10752#page=19.94

![Image](https://github.com/user-attachments/assets/9f8c6497-3bcb-4b91-a1c6-4c329d11c0c4)


> （2）The second doubt is, what is the dimension of batch['context']? Is it (batch_size, sequence_length, embedding_dim)? Before entering the LDM through cross-attention, does the context need additional modules for embedding?

The context tensor has dimensions $B \times L \times C$, where:

* $B$ is the batch size
* $L$ is the sequence length, which is always $L = 1$ since I concatenate the conditioning variables into a 1D embedding
* $C$ is the number of conditioning variables

The conditioning variables (rescaled to $[0,1]$) are passed directly to the U-Net’s cross-attention layer without any additional processing. I found this simple approach to work quite well.

**Comment by QijingGJ** [comment]
> Hi! I'm glad that you solved your issue!
> 
> > （1）The first doubt is whether the calculation of scale_factor uses only one sample. Why not all samples？
> 
> Read this section from the LDM paper https://arxiv.org/pdf/2112.10752#page=19.94
> 
> ![Image](https://github.com/user-attachments/assets/9f8c6497-3bcb-4b91-a1c6-4c329d11c0c4)
> 
> > （2）The second doubt is, what is the dimension of batch['context']? Is it (batch_size, sequence_length, embedding_dim)? Before entering the LDM through cross-attention, does the context need additional modules for embedding?
> 
> The context tensor has dimensions B × L × C , where:
> 
> * B
>    is the batch size
> * L
>    is the sequence length, which is always 
>     L
>     =
>     1
>    since I concatenate the conditioning variables into a 1D embedding
> * C
>    is the number of conditioning variables
> 
> The conditioning variables (rescaled to [ 0 , 1 ] ) are passed directly to the U-Net’s cross-attention layer without any additional processing. I found this simple approach to work quite well.

Oh, I understand. Thank you very much for your patient reply

**Comment by LemuelPuglisi** [comment]
No problem! Do not hesitate to contact me if you have other questions! My email address is lemuel.puglisi@phd.unict.it
